sns/parity_math.c

```diff
@@ -386,17 +746,84 @@ static void xor_calculate(struct m0_parity_math *math,
 		((uint8_t*)parity[0].b_addr)[ei] = pe;
 	}
 
+	return M0_RC(0);
 }
 
-static void reed_solomon_diff(struct m0_parity_math *math,
-			      struct m0_buf         *old,
-			      struct m0_buf         *new,
-			      struct m0_buf         *parity,
-			      uint32_t               index)
+#if ISAL_ENCODE_ENABLED
+static int isal_diff(struct m0_parity_math *math,
+		     struct m0_buf         *old,
+		     struct m0_buf         *new,
+		     struct m0_buf         *parity,
+		     uint32_t               index)
+{
+	struct m0_buf	diff_data_buf;
+	uint8_t	       *diff_data = NULL;
+	uint32_t	block_size;
+	uint32_t	alignment = sizeof(uint_fast32_t);
+	uint32_t	i;
+	int		ret = 0;
+
+	M0_ENTRY("math=%p, old=%p, new=%p, parity=%p, index=%u",
+		 math, old, new, parity, index);
+
+	M0_PRE(parity_math_invariant(math));
+	M0_PRE(old    != NULL);
+	M0_PRE(new    != NULL);
+	M0_PRE(parity != NULL);
+	M0_PRE(index  <  math->pmi_data_count);
+
+	if (old[index].b_nob != new[index].b_nob)
+		return M0_ERR_INFO(-EINVAL, "Data block size mismatch. "
+				   "Index=%u, Old data block size=%u, "
+				   "New data block size=%u", index,
+				   (uint32_t)old[index].b_nob,
+				   (uint32_t)new[index].b_nob);
+
+	block_size = new[index].b_nob;
+
+	/* It is assumed that the buffer size will always be a multiple of
+	 * 8-bytes (especially since block size currently is 4K) and this assert
+	 * is added with the hope that any deviation from this assumption is
+	 * caught during development instead of on the field.
+	 */
+	M0_ASSERT_INFO(m0_is_aligned(block_size, alignment),
+		       "block_size=%u is not %u-bytes aligned",
+		       block_size, alignment);
+
+	ALLOC_ARR_INFO(diff_data, block_size, "differential data block", ret);
+	if (diff_data == NULL)
+		return ret;
+
+	m0_buf_init(&diff_data_buf, diff_data, block_size);
+
+	/* Calculate differential data */
+	for (i = 0; i < (block_size / alignment); i++)
+		((uint_fast32_t *)diff_data)[i] =
+			((uint_fast32_t *)old[index].b_addr)[i] ^
+			((uint_fast32_t *)new[index].b_addr)[i];
+
+	/* Update differential parity */
+	ret = isal_encode_data_update(parity, &diff_data_buf, index,
+				      math->pmi_encode_tbls,
+				      math->pmi_data_count,
+				      math->pmi_parity_count);
+
+	m0_free(diff_data);
+	return M0_RC(ret);
+}
+#endif /* ISAL_ENCODE_ENABLED */
+
+#if !ISAL_ENCODE_ENABLED
+static int reed_solomon_diff(struct m0_parity_math *math,
+			     struct m0_buf         *old,
+			     struct m0_buf         *new,
+			     struct m0_buf         *parity,
+			     uint32_t               index)
 {
 	struct m0_matrix *mat;
-	uint32_t          ei;
-	uint32_t          ui;
+	uint32_t	  ei;
+	uint32_t	  ui;
+	uint32_t	  pi;
```

Comment:
Tabs are only used at the beginning of a line. "Internal" indentation is done with spaces. Please fix everywhere.